apps/btc_family/btc_txn.c

```diff
@@ -0,0 +1,288 @@
+ * @param which_request The expected request type enum
+ *
+ * @return bool Indicating if the check succeeded or failed
+ * @retval true If the query contains the expected request
+ * @retval false If the query does not contain the expected request
+ */
+static bool check_which_request(const btc_query_t *query,
+                                pb_size_t which_request);
+
+/**
+ * @brief Validates the derivation path received in the request from host
+ * @details The function validates the provided account derivation path in the
+ * request. If invalid path is detected, the function will send an error to the
+ * host and return false.
+ *
+ * @param request Reference to an instance of btc_sign_txn_request_t
+ * @return bool Indicating if the verification passed or failed
+ * @retval true If all the derivation path entries are valid
+ * @retval false If any of the derivation path entries are invalid
+ */
+static bool validate_request_data(const btc_sign_txn_request_t *request);
+
+/**
+ * @brief The function prepares and sends empty responses
+ *
+ * @param which_response Constant value for the response type to be sent
+ */
+static void send_response(pb_size_t which_response);
+
+/**
+ * @brief Takes already received and decoded query for the user confirmation.
+ * @details The function will verify if the query contains the BTC_SIGN_TXN type
+ * of request. Additionally, the wallet-id is validated for sanity and the
+ * derivation path for the account is also validated. After the validations,
+ * user is prompted about the action for confirmation. The function returns true
+ * indicating all the validation and user confirmation was a success. The
+ * function also duplicates the data from query into the g_txn_context for
+ * further processing.
+ *
+ * @param query Constant reference to the decoded query received from the host
+ *
+ * @return bool Indicating if the function actions succeeded or failed
+ * @retval true If all the validation and user confirmation was positive
+ * @retval false If any of the validation or user confirmation was negative
+ */
+static bool handle_initiate_query(const btc_query_t *query);
+
+/**
+ * @brief Handles fetching of the metadata/top-level transaction elements
+ * @details The function waits on USB event then decoding and validation of the
+ * received query. Post validation, based on the values in the query, the
+ * function allocates memory for storing input & output UTXOs in g_txn_context.
+ * Also, the data received in the query is duplicated into g_txn_context.
+ *
+ * @param query Reference to storage for decoding query from host
+ *
+ * @return bool Indicating if the function actions succeeded or failed
+ */
+static bool fetch_transaction_meta(btc_query_t *query);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+static txn_context *g_txn_context = NULL;
```

Comment:
Consider renaming to `btc_txn_context`.
`g_` comes from Hungarian notation and is disallowed in coding guidelines.

Reply:
changed. thanks

Reply:
resolving this and merging.